eclair-core/src/main/scala/fr/acinq/eclair/payment/send/OfferPayment.scala

```diff
@@ -0,0 +1,157 @@
+                      offer: Offer,
+                      amount: MilliSatoshi,
+                      quantity: Long,
+                      sendPaymentConfig: SendPaymentConfig) extends Command
+
+  case class WrappedMessageResponse(response: OnionMessageResponse) extends Command
+
+  case class SendPaymentConfig(externalId_opt: Option[String],
+                               maxAttempts: Int,
+                               routeParams: RouteParams,
+                               blocking: Boolean)
+
+  def apply(nodeParams: NodeParams,
+            postman: typed.ActorRef[Postman.Command],
+            paymentInitiator: ActorRef): Behavior[Command] = {
+    Behaviors.setup(context =>
+      Behaviors.receiveMessagePartial {
+        case PayOffer(replyTo, offer, amount, quantity, sendPaymentConfig) =>
+          if (!nodeParams.features.bolt12Features().areSupported(offer.features)) {
+            replyTo ! UnsupportedFeatures(offer.features.invoiceFeatures())
+            Behaviors.stopped
+          } else if (!offer.chains.contains(nodeParams.chainHash)) {
+            replyTo ! UnsupportedChains(offer.chains)
+            Behaviors.stopped
+          } else if (offer.expiry.exists(_ < TimestampSecond.now())) {
+            replyTo ! ExpiredOffer(offer.expiry.get)
+            Behaviors.stopped
+          } else if (offer.quantityMax.getOrElse(1L) < quantity) {
+            replyTo ! QuantityTooHigh(offer.quantityMax.getOrElse(1))
+            Behaviors.stopped
+          } else if (offer.amount.exists(_ * quantity > amount)) {
+            replyTo ! AmountInsufficient(offer.amount.get * quantity)
+            Behaviors.stopped
+          } else {
+            val payerKey = randomKey()
+            val request = InvoiceRequest(offer, amount, quantity, nodeParams.features.bolt12Features(), payerKey, nodeParams.chainHash)
+            sendInvoiceRequest(nodeParams, postman, paymentInitiator, context, request, payerKey, replyTo, nodeParams.onionMessageConfig.maxAttempts, sendPaymentConfig)
+          }
+      })
+  }
+
+  val rand = new Random
+
+  def sendInvoiceRequest(nodeParams: NodeParams,
+                         postman: typed.ActorRef[Postman.Command],
+                         paymentInitiator: ActorRef,
+                         context: ActorContext[Command],
+                         request: InvoiceRequest,
+                         payerKey: PrivateKey,
+                         replyTo: ActorRef,
+                         remainingAttempts: Int,
+                         sendPaymentConfig: SendPaymentConfig): Behavior[Command] = {
+    val destination = request.offer.contactInfo match {
+      case Left(blindedRoutes) =>
+        val blindedRoute = blindedRoutes(rand.nextInt(blindedRoutes.length))
+        OnionMessages.BlindedPath(blindedRoute)
+      case Right(nodeId) =>
+        OnionMessages.Recipient(nodeId, None, None)
+    }
+    // TODO: random choice of intermediate nodes
+    val intermediateNodes = Nil
```

Comment:
It would be nice to implement this in a follow-up PR, as directly connecting to the recipient (or the introduction node) leaks the payer's identity if it's the default behavior...what's your plan here? Making a path-finding call to the router?

Reply:
Right, that's something we'll need to add to the graph: we're missing feature bits associated to edges and/or vertices. That's something we'll also need when we do PTLCs, we'll need to know which edges can relay PTLC payments, so at some point we'll need to update our `Graph` data to contain those.